sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/ToNumberParser.scala

```diff
@@ -0,0 +1,577 @@
+          } else if (inputIndex < inputLength &&
+            inputBuffer(inputIndex) == MINUS_SIGN) {
+            negateResult = !negateResult
+            inputIndex += 1
+          } else {
+            // There is no plus or minus sign. Consume the token and remain at the same character in
+            // the input string.
+          }
+        case OptionalMinusSign() =>
+          if (inputIndex < inputLength &&
+            inputBuffer(inputIndex) == MINUS_SIGN) {
+            negateResult = !negateResult
+            inputIndex += 1
+          } else {
+            // There is no minus sign. Consume the token and remain at the same character in the
+            // input string.
+          }
+        case OpeningAngleBracket() =>
+          if (inputIndex >= inputLength ||
+            inputBuffer(inputIndex) != ANGLE_BRACKET_OPEN) {
+            // The input string did not contain an expected opening angle bracket.
+            return formatMatchFailure(input, originNumberFormat)
+          }
+          inputIndex += 1
+        case ClosingAngleBracket() =>
+          if (inputIndex >= inputLength ||
+            inputBuffer(inputIndex) != ANGLE_BRACKET_CLOSE) {
+            // The input string did not contain an expected closing angle bracket.
+            return formatMatchFailure(input, originNumberFormat)
+          }
+          negateResult = !negateResult
+          inputIndex += 1
+      }
+      formatIndex += 1
+    }
+    if (inputIndex < inputLength) {
+      // If we have consumed all the tokens in the format string, but characters remain unconsumed
+      // in the input string, then the input string does not match the format string.
+      formatMatchFailure(input, originNumberFormat)
+    } else {
+      getDecimal(negateResult)
+    }
+  }
+
+  /**
+   * Handle parsing the input string for the given expected DigitGroups from the format string.
+   *
+   * @param digitGroups the expected DigitGroups from the format string
+   * @param inputBuffer the input string provided to the original parsing method
+   * @param startingInputIndex the input index within the input string to begin parsing here
+   * @param reachedDecimalPoint true if we have already parsed past the decimal point
+   * @return the new updated index within the input string to resume parsing, or None on error
+   */
+  private def parseDigitGroups(
+      digitGroups: DigitGroups,
+      inputBuffer: Array[Byte],
+      startingInputIndex: Int,
+      reachedDecimalPoint: Boolean): Option[Int] = {
+    val expectedDigits: Seq[Digits] = digitGroups.digits
+    val inputLength = inputBuffer.length
+    // Consume characters from the current input index forwards in the input string as long as
+    // they are digits (0-9) or the thousands separator (,).
+    var numDigits = 0
+    var inputIndex = startingInputIndex
+    actualDigitLengths.clear()
+    while (inputIndex < inputLength &&
+      (inputBuffer(inputIndex).toChar match {
```

Comment:
can we create a method for it? It's weird to put so much code in the while condition.

Reply:
Sure, good idea, done!